core/git.go

```diff
@@ -141,111 +132,329 @@ func (repo *RemoteGitRepository) PBDefinitions(ctx context.Context) ([]*pb.Defin
 	return nil, nil
 }
 
-func (repo *RemoteGitRepository) Head(ctx context.Context) (GitRefBackend, error) {
-	return &RemoteGitRef{
+func (repo *RemoteGitRepository) Ref(ctx context.Context, refstr string) (GitRefBackend, error) {
+	ref := &RemoteGitRef{
 		Query: repo.Query,
 		Repo:  repo,
-	}, nil
-}
+	}
 
-func (repo *RemoteGitRepository) Ref(ctx context.Context, ref string) (GitRefBackend, error) {
-	return &RemoteGitRef{
-		Query: repo.Query,
-		Repo:  repo,
-		Ref:   ref,
-	}, nil
+	// force resolution now, since the remote might change, and we don't want inconsistencies
+	var err error
+	ref.Commit, ref.FullRef, err = ref.resolve(ctx, refstr)
+	if err != nil {
+		return nil, err
+	}
+
+	return ref, nil
 }
 
 func (repo *RemoteGitRepository) Tags(ctx context.Context, patterns []string) ([]string, error) {
-	// standardize to the same ref that goes into the state (see llb.Git)
-	remote, err := gitutil.ParseURL(repo.URL)
-	if errors.Is(err, gitutil.ErrUnknownProtocol) {
-		remote, err = gitutil.ParseURL("https://" + repo.URL)
+	svcs, err := repo.Query.Services(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get services: %w", err)
 	}
+	detach, _, err := svcs.StartBindings(ctx, repo.Services)
 	if err != nil {
 		return nil, err
 	}
+	defer detach()
 
 	queryArgs := []string{
 		"ls-remote",
 		"--tags", // we only want tags
 		"--refs", // we don't want to include ^{} entries for annotated tags
-		remote.Remote,
+		repo.URL.Remote(),
 	}
 	if len(patterns) > 0 {
 		queryArgs = append(queryArgs, "--")
 		queryArgs = append(queryArgs, patterns...)
 	}
-	cmd := exec.CommandContext(ctx, "git", queryArgs...)
+	git, cleanup, err := repo.setup(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer cleanup()
+	out, err := git.Run(ctx, queryArgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	tags := []string{}
+	scanner := bufio.NewScanner(bytes.NewReader(out))
+	for scanner.Scan() {
+		fields := strings.Fields(scanner.Text())
+		if len(fields) < 2 {
+			continue
+		}
+
+		// this API is to fetch tags, not refs, so we can drop the `refs/tags/`
+		// prefix
+		tag := strings.TrimPrefix(fields[1], "refs/tags/")
+
+		tags = append(tags, tag)
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("error scanning git output: %w", err)
+	}
+
+	return tags, nil
+}
+
+func (repo *RemoteGitRepository) setup(ctx context.Context) (_ *gitutil.GitCLI, _ func() error, rerr error) {
+	var opts []gitutil.Option
+
+	cleanups := buildkit.Cleanups{}
+	defer func() {
+		if rerr != nil {
+			cleanups.Run()
+		}
+	}()
+
+	if repo.AuthToken.Self != nil {
+		username := "x-access-token"
+		if repo.URL.Host == "bitbucket.org" {
+			// NOTE: bitbucket.org is picky, and needs *this* username
+			username = "x-token-auth"
+		}
```

Comment:
cc @grouville, this is a simplification of part of #8778 - it turns out, we don't need the whole credential helper here?

Anyways, just wanted to check that this bit seemed good to you, all our tests seem to pass correctly :tada: